src/app.js

```diff
@@ -25,107 +25,92 @@ export function app(props) {
     Object.keys(withActions || {}).map(function(name) {
-      return typeof withActions[name] === "function"
-        ? (actions[name] = function(data) {
-            return typeof (data = withActions[name](
-              getPath(lastPath, appState),
-              getPath(lastPath, appActions),
-              data
-            )) === "function"
-              ? data(update)
-              : update(data)
-          })
-        : createActions(
-            actions[name] || (actions[name] = {}),
-            withActions[name],
-            lastPath.concat(name)
-          )
+      if (typeof withActions[name] === "function") {
+        actions[name] = function(data) {
+          return typeof (data = withActions[name](withState, actions, data)) ===
+          "function"
+            ? data(update)
+            : update(data)
+        }
+      } else {
+        setupActions(
+          (actions[name] = {}),
+          withState[name] || (withState[name] = {}),
+          withActions[name]
+        )
+      }
     })
 
-    function update(withState) {
-      if (typeof withState === "function") {
-        return update(withState(getPath(lastPath, appState)))
-      }
-      if (
-        withState &&
-        (withState = setPath(
-          lastPath,
-          merge(getPath(lastPath, appState), withState),
-          appState
-        ))
-      ) {
-        requestRender((appState = withState))
-      }
-      return appState
+    function update(data) {
+      return (
+        typeof data === "function"
+          ? update(data(withState))
+          : data && requestRender(copy(withState, data)),
+        withState
+      )
     }
   }
 
-  function set(prop, value, source) {
-    var target = merge(source)
-    target[prop] = value
+  function copy(target, source) {
+    for (var i in source) {
+      target[i] = source[i]
+    }
     return target
   }
 
-  function getPath(paths, source) {
-    return paths.length === 0
-      ? source
-      : source && getPath(paths.slice(1), source[paths[0]])
-  }
-
-  function setPath(paths, value, source) {
-    var name = paths[0]
-    return paths.length === 0
-      ? value
-      : set(
-          name,
-          paths.length > 1
-            ? setPath(
-                paths.slice(1),
-                value,
-                source && name in source ? source[name] : {}
-              )
-            : value,
-          source
-        )
+  function merge(target, source, result) {
+    return copy(copy((result = {}), target), source)
   }
 
-  function merge(target, source) {
-    var result = {}
-    for (var i in target) {
-      result[i] = target[i]
-    }
-    for (var i in source) {
-      result[i] = source[i]
-    }
-    return result
+  function hydrate(element, map) {
+    return (
+      element &&
+      h(
+        element.tagName.toLowerCase(),
+        {},
+        map.call(element.childNodes, function(element) {
+          return element.nodeType === 3
```

Comment:
I'm guessing we can't use `Node.TEXT_NODE` here because uglify won't know to replace it with the constant? I also thought we used to have a comment here about the meaning of this magic number.

Reply:
Could also copy what Preact does, and check if `element.splitText !== undefined`.   In a conference talk, `developit` mentioned the perf win was substantial, since `nodeType` is a getter function.

https://youtu.be/LY6y3HbDVmg?t=19m53s
https://github.com/developit/preact/blob/master/src/vdom/diff.js#L76